In [59]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/ultimate-spotify-tracks-db/SpotifyFeatures.csv


### Loading the Dataset

The Spotify dataset is loaded into a Pandas DataFrame.  
This dataset contains track-level metadata and audio features required for building a content-based recommendation system.  
Initial inspection ensures the dataset is correctly loaded.


In [60]:
df=pd.read_csv("/kaggle/input/ultimate-spotify-tracks-db/SpotifyFeatures.csv")

### Dataset Inspection

Displays the first few rows of the dataset to understand feature structure, column names, and data types.  
This step helps verify that all relevant attributes are present before preprocessing.


In [61]:
df.head()

genre        artist_name                        track_name  \
0  Movie     Henri Salvador       C'est beau de faire un Show   
1  Movie  Martin & les fées  Perdu d'avance (par Gad Elmaleh)   
2  Movie    Joseph Williams    Don't Let Me Be Lonely Tonight   
3  Movie     Henri Salvador    Dis-moi Monsieur Gordon Cooper   
4  Movie       Fabien Nataf                         Ouverture   

                 track_id  popularity  acousticness  danceability  \
0  0BRjO6ga9RKCKjfDqeFgWV           0         0.611         0.389   
1  0BjC1NfoEOOusryehmNudP           1         0.246         0.590   
2  0CoSDzoNIKCRs124s9uTVy           3         0.952         0.663   
3  0Gc6TVm52BwZD07Ki6tIvf           0         0.703         0.240   
4  0IuslXpMROHdEPvSl1fTQK           4         0.950         0.331   

   duration_ms  energy  instrumentalness key  liveness  loudness   mode  \
0        99373   0.910             0.000  C#    0.3460    -1.828  Major   
1       137373   0.737             0.000  F#    0.1510    -5.559  Minor   
2       170267   0.131             0.000   C    0.1030   -13.879  Minor   
3       152427   0.326             0.000  C#    0.0985   -12.178  Major   
4        82625   0.225             0.123   F    0.2020   -21.150  Major   

   speechiness    tempo time_signature  valence  
0       0.0525  166.969            4/4    0.814  
1       0.0868  174.003            4/4    0.816  
2       0.0362   99.488            5/4    0.368  
3       0.0395  171.758            4/4    0.227  
4       0.0456  140.576            4/4    0.390

In [62]:
df.tail()

genre               artist_name                      track_name  \
232720  Soul                     Slave                    Son Of Slide   
232721  Soul  Jr Thomas & The Volcanos                    Burning Fire   
232722  Soul              Muddy Waters  (I'm Your) Hoochie Coochie Man   
232723  Soul                   R.LUM.R                   With My Words   
232724  Soul            Mint Condition  You Don't Have To Hurt No More   

                      track_id  popularity  acousticness  danceability  \
232720  2XGLdVl7lGeq8ksM6Al7jT          39       0.00384         0.687   
232721  1qWZdkBl4UVPj9lK6HuuFM          38       0.03290         0.785   
232722  2ziWXUmQLrXTiYjCg2fZ2t          47       0.90100         0.517   
232723  6EFsue2YbIG4Qkq8Zr9Rir          44       0.26200         0.745   
232724  34XO9RwPMKjbvRry54QzWn          35       0.09730         0.758   

        duration_ms  energy  instrumentalness key  liveness  loudness   mode  \
232720       326240   0.714          0.544000   D    0.0845   -10.626  Major   
232721       282447   0.683          0.000880   E    0.2370    -6.944  Minor   
232722       166960   0.419          0.000000   D    0.0945    -8.282  Major   
232723       222442   0.704          0.000000   A    0.3330    -7.137  Major   
232724       323027   0.470          0.000049  G#    0.0836    -6.708  Minor   

        speechiness    tempo time_signature  valence  
232720       0.0316  115.542            4/4    0.962  
232721       0.0337  113.830            4/4    0.969  
232722       0.1480   84.135            4/4    0.813  
232723       0.1460  100.031            4/4    0.489  
232724       0.0287  113.897            4/4    0.479

### Dataset Dimensions

Checks the number of rows and columns in the dataset.  
This gives an overview of dataset size and helps estimate computational requirements.


In [63]:
df.shape

(232725, 18)

# Basic Analysis

In [64]:
df.dtypes

genre                object
artist_name          object
track_name           object
track_id             object
popularity            int64
acousticness        float64
danceability        float64
duration_ms           int64
energy              float64
instrumentalness    float64
key                  object
liveness            float64
loudness            float64
mode                 object
speechiness         float64
tempo               float64
time_signature       object
valence             float64
dtype: object

In [65]:
df.describe()

popularity   acousticness   danceability   duration_ms  \
count  232725.000000  232725.000000  232725.000000  2.327250e+05   
mean       41.127502       0.368560       0.554364  2.351223e+05   
std        18.189948       0.354768       0.185608  1.189359e+05   
min         0.000000       0.000000       0.056900  1.538700e+04   
25%        29.000000       0.037600       0.435000  1.828570e+05   
50%        43.000000       0.232000       0.571000  2.204270e+05   
75%        55.000000       0.722000       0.692000  2.657680e+05   
max       100.000000       0.996000       0.989000  5.552917e+06   

              energy  instrumentalness       liveness       loudness  \
count  232725.000000     232725.000000  232725.000000  232725.000000   
mean        0.570958          0.148301       0.215009      -9.569885   
std         0.263456          0.302768       0.198273       5.998204   
min         0.000020          0.000000       0.009670     -52.457000   
25%         0.385000          0.000000       0.097400     -11.771000   
50%         0.605000          0.000044       0.128000      -7.762000   
75%         0.787000          0.035800       0.264000      -5.501000   
max         0.999000          0.999000       1.000000       3.744000   

         speechiness          tempo        valence  
count  232725.000000  232725.000000  232725.000000  
mean        0.120765     117.666585       0.454917  
std         0.185518      30.898907       0.260065  
min         0.022200      30.379000       0.000000  
25%         0.036700      92.959000       0.237000  
50%         0.050100     115.778000       0.444000  
75%         0.105000     139.054000       0.660000  
max         0.967000     242.903000       1.000000

In [66]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 232725 entries, 0 to 232724
Data columns (total 18 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   genre             232725 non-null  object 
 1   artist_name       232725 non-null  object 
 2   track_name        232724 non-null  object 
 3   track_id          232725 non-null  object 
 4   popularity        232725 non-null  int64  
 5   acousticness      232725 non-null  float64
 6   danceability      232725 non-null  float64
 7   duration_ms       232725 non-null  int64  
 8   energy            232725 non-null  float64
 9   instrumentalness  232725 non-null  float64
 10  key               232725 non-null  object 
 11  liveness          232725 non-null  float64
 12  loudness          232725 non-null  float64
 13  mode              232725 non-null  object 
 14  speechiness       232725 non-null  float64
 15  tempo             232725 non-null  float64
 16  time_signature    23

Converting ms to seconds for ease

In [67]:
df['duration_ms']=round(df['duration_ms']/1000,2)
df.rename(columns={'duration_ms':'duration_sec'},inplace=True)
df.head()

genre        artist_name                        track_name  \
0  Movie     Henri Salvador       C'est beau de faire un Show   
1  Movie  Martin & les fées  Perdu d'avance (par Gad Elmaleh)   
2  Movie    Joseph Williams    Don't Let Me Be Lonely Tonight   
3  Movie     Henri Salvador    Dis-moi Monsieur Gordon Cooper   
4  Movie       Fabien Nataf                         Ouverture   

                 track_id  popularity  acousticness  danceability  \
0  0BRjO6ga9RKCKjfDqeFgWV           0         0.611         0.389   
1  0BjC1NfoEOOusryehmNudP           1         0.246         0.590   
2  0CoSDzoNIKCRs124s9uTVy           3         0.952         0.663   
3  0Gc6TVm52BwZD07Ki6tIvf           0         0.703         0.240   
4  0IuslXpMROHdEPvSl1fTQK           4         0.950         0.331   

   duration_sec  energy  instrumentalness key  liveness  loudness   mode  \
0         99.37   0.910             0.000  C#    0.3460    -1.828  Major   
1        137.37   0.737             0.000  F#    0.1510    -5.559  Minor   
2        170.27   0.131             0.000   C    0.1030   -13.879  Minor   
3        152.43   0.326             0.000  C#    0.0985   -12.178  Major   
4         82.62   0.225             0.123   F    0.2020   -21.150  Major   

   speechiness    tempo time_signature  valence  
0       0.0525  166.969            4/4    0.814  
1       0.0868  174.003            4/4    0.816  
2       0.0362   99.488            5/4    0.368  
3       0.0395  171.758            4/4    0.227  
4       0.0456  140.576            4/4    0.390

### Missing Value Analysis

Identifies null or missing values across all columns.  
Ensures data quality before applying transformations and similarity computations.
Also checking if duplicates exist.

In [68]:
df.isnull().sum()

genre               0
artist_name         0
track_name          1
track_id            0
popularity          0
acousticness        0
danceability        0
duration_sec        0
energy              0
instrumentalness    0
key                 0
liveness            0
loudness            0
mode                0
speechiness         0
tempo               0
time_signature      0
valence             0
dtype: int64

In [69]:
df.dropna(inplace=True)

In [70]:
df.isnull().sum()

genre               0
artist_name         0
track_name          0
track_id            0
popularity          0
acousticness        0
danceability        0
duration_sec        0
energy              0
instrumentalness    0
key                 0
liveness            0
loudness            0
mode                0
speechiness         0
tempo               0
time_signature      0
valence             0
dtype: int64

In [71]:
df.duplicated().sum()

np.int64(0)


splitting up the categorical and numerical columns

In [72]:
cat_cols=df.select_dtypes(include=object).columns
num_cols=df.select_dtypes(include=np.number).columns

In [73]:
for col in cat_cols:
    print(df[col].value_counts())

genre
Comedy              9681
Soundtrack          9646
Indie               9543
Jazz                9441
Pop                 9386
Electronic          9377
Children’s Music    9353
Folk                9299
Hip-Hop             9295
Rock                9272
Alternative         9263
Classical           9256
Rap                 9232
World               9095
Soul                9089
Blues               9023
R&B                 8992
Anime               8936
Reggaeton           8927
Ska                 8874
Reggae              8771
Dance               8701
Country             8664
Opera               8280
Movie               7806
Children's Music    5403
A Capella            119
Name: count, dtype: int64
artist_name
Giuseppe Verdi            1394
Giacomo Puccini           1137
Kimbo Children's Music     971
Nobuo Uematsu              825
Richard Wagner             804
                          ... 
Major Harris                 1
The Magnolia                 1
Scienze                      1
De

### Aggregation

aggregating track ids together to prevent multiple entries and biasness

In [ ]:
df = df.groupby('track_id', as_index=False).agg({
    'track_name': 'first',
    'artist_name': 'first',
    'genre': 'first',
    'popularity': 'mean',
    'acousticness': 'mean',
    'danceability': 'mean',
    'energy': 'mean',
    'instrumentalness': 'mean',
    'liveness': 'mean',
    'loudness': 'mean',
    'speechiness': 'mean',
    'tempo': 'mean',
    'valence': 'mean',
    'key': lambda x: x.mode()[0],
    'mode': lambda x: x.mode()[0],
    'time_signature': lambda x: x.mode()[0]
})

In [ ]:
df['track_id'].value_counts()

### Extracting meta data

These features dont contribute in building the recommendation system.
So separating features like track names, artist names, and other descriptive metadata.  
This metadata is later used to display human-readable recommendations.

In [ ]:
meta=df[['track_id','track_name','artist_name']]

### Removing Non-Useful Columns from main dataframe

Removes columns that do not contribute to content-based similarity, such as identifiers or redundant metadata.  
This simplifies the feature space.


In [ ]:
df.drop(['track_id','track_name','artist_name'],axis=1,inplace=True)

In [ ]:
cat_cols=df.select_dtypes(include=object).columns
num_cols=df.select_dtypes(include=np.number).columns

In [ ]:
for col in cat_cols:
    print(df[col].value_counts())

# Preprocessing

### Encoding Categorical Features

Encodes categorical attributes such as artists or genres into numerical form if required.  
Ensures compatibility with mathematical similarity methods.
Made use of ***One Hot Encoding*** because number of classes or not very high.

In [ ]:
from sklearn.preprocessing import OneHotEncoder

ohe=OneHotEncoder(sparse_output=False)
cat_cols_encoded=ohe.fit_transform(df[cat_cols])
encoded_cols=ohe.get_feature_names_out()
cat_cols_encoded=pd.DataFrame(cat_cols_encoded,columns=encoded_cols,index=df.index)
cat_cols_encoded

In [ ]:
df[num_cols]

### Checking for skewness in numerical features

Checking if numerical columns are skewed, so that appropriate transformation can be applied.

In [ ]:
df[num_cols].skew()

Seperating skewed columns, i.e, the columns with absolute skew value greater than 1

In [ ]:
skewed_cols=[]
for col in num_cols:
    if abs(df[col].skew())>1:
        skewed_cols.append(col)
skewed_cols

### Transformation

Used automated power transformer class from scikit learn library with yeo johnson method, so that it can handle positive, negative and zero values efficiently and transform them.

In [ ]:
from sklearn.preprocessing import PowerTransformer

pt=PowerTransformer(method='yeo-johnson')
df_transformed=pt.fit_transform(df[skewed_cols])
df_transformed=pd.DataFrame(df_transformed,columns=skewed_cols,index=df[num_cols].index)
df_transformed.head()

Seperating non skewed columns 

In [ ]:
non_skewed=set(num_cols)-set(skewed_cols)
non_skewed=list(non_skewed)

### Final numerical data preparation

In [ ]:
num_data=pd.concat([df_transformed,df[non_skewed]],axis=1)

### Feature Scaling

Applies normalization/standardization to ensure all audio features are on a comparable scale.  
This prevents features with large ranges from dominating similarity calculations.

Used **MinMax Scaler** because recommendation systems work based on distances.

In [ ]:
from sklearn.preprocessing import MinMaxScaler

mms=MinMaxScaler()
num_data_scaled=mms.fit_transform(num_data)
num_data_scaled=pd.DataFrame(num_data_scaled,columns=num_data.columns,index=num_data.index)


### Final Feature Dataset Assembly

Merges numerical audio features and encoded categorical data into a single dataset.  
This dataset fully represents each song for recommendation purposes.


In [ ]:
df_final=pd.concat([cat_cols_encoded,num_data_scaled],axis=1)
df_final.tail()

# Model Training

### Sparse Matrix Optimization

Converts the final feature matrix into a sparse format.  
This reduces memory usage and improves performance during similarity computations.


In [ ]:
from scipy.sparse import csr_matrix
df_final_csr=csr_matrix(df_final)

### Similarity Metric Setup

Prepares the dataset for cosine similarity calculations.  
Cosine similarity measures how similar two songs are based on feature direction rather than magnitude.
This function outputs the **Cosine Similarity matrix** with which the model picks the suggested songs.

In [ ]:
meta[meta['track_name']=='Houdini']

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity
def recommendation_matrix(song,artist,songs_data):
    song_row=meta[(meta['track_name'].str.lower()==song.lower()) & (meta['artist_name'].str.lower()==artist.lower())]
    song_idx=song_row.index[0]
    if not song_row.empty:
        similarity_matrix=cosine_similarity(songs_data[song_idx],songs_data).flatten()
        return similarity_matrix,song_idx
    return "Song doesnt Exist",-1
      

### Recommendation Logic

Defines the core logic for identifying similar songs based on cosine similarity scores.  
The model ranks songs and excludes the input song itself from results.


### Content-Based Similarity Model

Constructs the similarity model using cosine similarity across all tracks.  
This model forms the core of the recommendation system and consists of two functions for two seperate purposes.

The ***content-based*** function returns songs similar in audio features, while the ***artist-based function*** ensures recommendations are restricted to the same artist. For example, querying ‘Night Changes’ produces top recommendations from One Direction in the artist-based approach, while the content-based function may include similar songs from other artists. This dual approach allows flexibility for users seeking either same-artist recommendations or similar-sounding songs. Therecommendation results are formatted into a clean, user-readable list.  
Improves presentation and usability.


#### Content Based Function

In [ ]:
def top5_recommendations(song,artist,songs_data):
    final_show=[]
    similar,idx=recommendation_matrix(song,artist,songs_data)
    if isinstance(similar,str):
        return "Song Not Found"
    top5=similar.argsort()[::-1]
    top5=top5[top5!=idx][:5]
    top5_songs=meta.iloc[top5][['track_name','artist_name']].values
    for song in top5_songs:
        final_show.append(song[0]+" "+"by"+" "+song[1])
    return final_show,top5


#### Artist Based Function

In [ ]:
def recommend_top5_song_by_artist(song, artist,songs_data, meta):
    final_show=[]
    similar,idx=recommendation_matrix(song,artist,songs_data)
    if isinstance(similar,str):
        return "Song Not Found"
    same_artist_mask = (meta['artist_name'].str.lower().values==artist.lower())
    similar[~same_artist_mask] = -1
    top5=similar.argsort()[::-1][1:6]
    top5=top5[top5!=idx][:5]
    top5_songs=meta.iloc[top5][['track_name','artist_name']].values
    for song in top5_songs:
        final_show.append(song[0]+" "+"by"+" "+song[1])
    return final_show,top5

# Test Case Validation

### Test Case: Song-Based Recommendation

Tests the recommendation system using a popular song as input.  
Verifies similarity accuracy and output format.


In [ ]:
print(top5_recommendations("perfect","ed sheeran",df_final_csr)[0])

### Test Case: Artist-Aware Recommendation

Tests recommendations when both song name and artist are provided.  
Confirms correct disambiguation and relevance.


In [ ]:
print(recommend_top5_song_by_artist("IDGAF","Dua lipa",df_final_csr,meta)[0])

# Evaluation

### Baseline Comparison

Baseline comparison is used to verify that the recommendation system performs better than random chance. The average cosine similarity of model-generated recommendations is compared against randomly selected songs. Self-similarity is explicitly removed to ensure fair evaluation. Higher similarity scores than the baseline confirm meaningful recommendations.


In [ ]:
def avg_similarity(idx, rec_indices, sim_matrix):
    if isinstance(rec_indices, (int, np.integer)):
        rec_indices = [rec_indices]      
    rec_indices = [i for i in rec_indices if i != idx]
    return np.mean(sim_matrix[rec_indices])

evaluating content based model

In [ ]:
import random 
random_recs = random.sample(range(df_final_csr.shape[0]), 10) 
sim_matrix1,idx1=recommendation_matrix("Perfect","ed sheeran",df_final_csr) 
avg_model=avg_similarity(idx1,top5_recommendations("Perfect","ed sheeran",df_final_csr)[1],sim_matrix1)  
avg_random=avg_similarity(idx1,random_recs,sim_matrix1) 
print(avg_model,avg_random)

evaluating artist based model

In [ ]:
sim_matrix2,idx2=recommendation_matrix("Perfect","ed sheeran",df_final_csr) 
avg_model2=avg_similarity(idx2,top5_recommendations("IDGAF","Dua lipa",df_final_csr)[1],sim_matrix2)  
avg_random=avg_similarity(idx2,random_recs,sim_matrix2) 
print(avg_model2,avg_random)

Content-based recommendations show a very high similarity (0.972) compared to the random baseline (0.391), indicating strong feature alignment. Artist-based recommendations improve over random as well (0.455 vs 0.391) but are noticeably weaker than full content-based similarity, which is still valid because there is no need an artist performs songs with similar music features.

### Leave One Out Strategy Test

This evaluation uses a leave-one-out strategy where, for each sampled song, the average cosine similarity of its top-k recommendations is compared against a random baseline. By repeating this across multiple songs and averaging the scores, the method verifies that the model consistently produces more similar recommendations than chance.

In [ ]:
def evaluate_model(song_indices, k=5):
    model_scores = []
    random_scores = []
    for idx in song_indices:
        sim = cosine_similarity(df_final_csr[idx], df_final_csr).flatten()
        topk = sim.argsort()[::-1]
        topk = [i for i in topk if i != idx][:k]
        randk = random.sample(
            [i for i in range(len(sim)) if i != idx],
            k
        )
        model_scores.append(np.mean(sim[topk]))
        random_scores.append(np.mean(sim[randk]))
    return np.mean(model_scores), np.mean(random_scores)

In [ ]:
sample_indices = random.sample(range(df_final_csr.shape[0]), 50)
evaluate_model(sample_indices)

This score validates that the recommendation system computes correct cosine similarities and the recommendations provided are valid when compared to a random suggestor.

### Notebook Completion

Marks the end of the notebook.  
All components of the content-based Spotify recommendation system have been successfully implemented and tested.
